public/app/core/utils/EventsProvider.tsx

```diff
@@ -0,0 +1,101 @@
+import React, { ComponentType, FunctionComponent, useEffect } from 'react';
+import { Observable, Subject, Unsubscribable } from 'rxjs';
+import { filter } from 'rxjs/operators';
+import { AppEvent, Subtract } from '@grafana/data';
+
+import appEvents from '../app_events';
+
+export interface EventsContextApi<SubscribeEvents, PublishEvents> {
+  events: (originFilter: string) => Observable<AppEvent<SubscribeEvents>>;
+  publish: (event: AppEvent<PublishEvents>, origin: string, payload?: PublishEvents) => void;
+}
+
+export interface RootEventsContextApi {
+  rootEvents: Observable<AppEvent<any>>;
+  rootPublish: (event: AppEvent<any>, payload?: any) => void;
+}
+
+const events = new Subject<AppEvent<any>>();
+const publish = <T extends {} = {}>(event: AppEvent<T>, payload?: T) => {
+  events.next({ ...event, payload });
+  appEvents.emit(event, payload);
+};
+
+export const rootEventing: RootEventsContextApi = {
+  rootEvents: events,
+  rootPublish: publish,
+};
+
+export const EventsContext = React.createContext(rootEventing);
+EventsContext.displayName = 'EventsContext';
+
+export const EventsProvider = ({ children }: { children: React.ReactNode }) => {
+  return <EventsContext.Provider value={rootEventing}>{children}</EventsContext.Provider>;
+};
+
+export const provideEvents = (component: ComponentType<any>) => (props: any) => (
+  <EventsProvider>{React.createElement(component, { ...props })}</EventsProvider>
+);
+
+export const withEvents = <SubscribeEvents extends {}, PublishEvents extends {}>(
+  ...subscribeEvents: Array<AppEvent<SubscribeEvents>>
+) => {
+  type ComponentWithEventingType = EventsContextApi<SubscribeEvents, PublishEvents>;
+  return <C extends ComponentWithEventingType>(Component: ComponentType<C>) => {
+    type ComponentWithOutEventingType = Subtract<C, ComponentWithEventingType>;
+
+    const ComponentWithEventing: FunctionComponent<ComponentWithOutEventingType> = (props: any) => {
+      useEffect(() => {
+        return function unsubscribe() {
+          if (subscription) {
+            subscription.unsubscribe();
+            subscription = null;
+          }
+          events.unsubscribe();
+        };
+      }, []);
+      let subscription: Unsubscribable | null = null;
+      const events = new Subject<AppEvent<SubscribeEvents>>();
+      const subscribe = (origin: string): Observable<AppEvent<SubscribeEvents>> => {
+        subscription = props.rootEvents
+          .pipe(
+            filter((event: AppEvent<any>) => {
+              if (event.origin === origin) {
+                return false;
+              }
+              const found = subscribeEvents.filter(value => value.name === event.name)[0];
+              return found !== undefined;
+            })
+          )
+          .subscribe((value: AppEvent<SubscribeEvents>) => {
+            if (events) {
+              events.next(value);
+            }
+          });
+
+        return events.asObservable();
```

Comment:
This feels like it will leak if the component won't unsubscribe. This HOC will subscribe and unsubscribe but the same needs to be done in the component anyway doesn't it.

Reply:
No, it doesn't. When the Subject is unsubscribed all observables are unsubscribes as well AFAIK.

Reply:
> When the Subject is unsubscribed all observables are unsubscribes as well AFAIK.

Not sure about that and not sure that applies here. If I follow the code correctly:
- `events` is subject to which we just push values
- we subscribe to `rootEvents` to push the values to `events` this creates `subscription1`
- we return `events` as observable and component will subscribe to it creating `subscription2`
- when we unmount we unsubscribe from `subscription1`, `rootEvents` still produce values but there is nothing to push the values into `events`.
- `events` though does not know about that and `subscription2` still exists if component did not manually unsubscribe so even though the events should not flow the `subscription2` should keep the `events` subject alive creating memory leak.

Reply:
I'll have to test this but I think the only leak is in the rootEvents right now. I'll do some more testing and come back.